components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt

```diff
@@ -318,9 +328,55 @@ class SessionManager(
      * Informs this [SessionManager] that the OS is in low memory condition so it
      * can reduce its allocated objects.
      */
+    @Deprecated("Use onTrimMemory instead.", replaceWith = ReplaceWith("onTrimMemory"))
     fun onLowMemory() {
-        delegate.onLowMemory()
-        store?.syncDispatch(SystemAction.LowMemoryAction)
+        onTrimMemory(ComponentCallbacks2.TRIM_MEMORY_MODERATE)
+    }
+
+    @Synchronized
+    @Suppress("NestedBlockDepth")
+    override fun onTrimMemory(level: Int) {
+        val clearThumbnails = level >= ComponentCallbacks2.TRIM_MEMORY_RUNNING_LOW
+        val closeEngineSessions = level >= ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL
+
+        logger.debug("onTrimMemory($level): clearThumbnails=$clearThumbnails, closeEngineSessions=$closeEngineSessions")
+
+        if (!clearThumbnails && !closeEngineSessions) {
+            // Nothing to do for now.
+            return
+        }
+
+        val selectedSession = selectedSession
+        val states = mutableMapOf<String, EngineSessionState>()
+        val sessions = sessions
+
+        var clearedThumbnails = 0
+        var unlinkedEngineSessions = 0
+
+        sessions.forEach {
+            if (it != selectedSession) {
+                if (clearThumbnails) {
+                    it.thumbnail = null
+                    clearedThumbnails++
+                }
+
+                if (closeEngineSessions) {
+                    val state = it.engineSessionHolder.engineSession?.saveState()
+                    linker.unlink(it)
+
+                    if (state != null) {
+                        it.engineSessionHolder.engineSessionState = state
+                        states[it.id] = state
+                    }
```

Comment:
We're doing this on `TRIM_MEMORY_RUNNING_CRITICAL,` but presumably `saveState` is also requiring some memory. Just wondering if we might fail to do this because there's not enough memory left in which case `TRIM_MEMORY_RUNNING_LOW` would be better for clearing engine sessions as well? 🤔 Of course, that would happen more often then. Just a thought. +1 for landing as is.

Reply:
You are right in theory, but: The implementation of `saveState` has changed in the past. Nowadays `GeckoEngineSession` always has a reference to the last state (since GV pushes it to us) and `saveState()` only returns that reference :)